File: packages/react-native/android/src/main/java/com/hotupdater/SignatureVerifier.kt
package com.hotupdater

import android.content.Context
import android.util.Base64
import android.util.Log
import java.io.File
import java.security.KeyFactory
import java.security.PublicKey
import java.security.Signature
import java.security.spec.X509EncodedKeySpec

/**
 * Prefix for signed file hash format.
 */
private const val SIGNED_HASH_PREFIX = "sig:"

/**
 * Custom exceptions for signature verification errors.
 *
 * **IMPORTANT**: The error messages in these exceptions are used by the JavaScript layer
 * (`packages/react-native/src/types.ts`) to detect signature verification failures.
 * If you change these messages, update `isSignatureVerificationError()` in types.ts accordingly.
 */
sealed class SignatureVerificationException(
    message: String,
) : Exception(message) {
    class PublicKeyNotConfigured :
        SignatureVerificationException(
            "Public key not configured for signature verification. " +
                "Add 'hot_updater_public_key' to res/values/strings.xml",
        )

    class InvalidPublicKeyFormat :
        SignatureVerificationException(
            "Public key format is invalid. Ensure the public key is in PEM format (BEGIN PUBLIC KEY)",
        )

    class MissingFileHash :
        SignatureVerificationException(
            "File hash is missing or empty. Ensure the bundle update includes a valid file hash",
        )

    class InvalidSignatureFormat :
        SignatureVerificationException(
            "Signature format is invalid or corrupted. The signature data is malformed or cannot be decoded",
        )

    class SignatureVerificationFailed :
        SignatureVerificationException(
            "Bundle signature verification failed. The bundle may be corrupted or tampered with",
        )

    class FileHashMismatch :
        SignatureVerificationException(
            "File hash verification failed. The bundle file hash does not match the expected value. File may be corrupted",
        )

    class FileReadFailed :
        SignatureVerificationException(
            "Failed to read file for verification. Could not read file for hash verification",
        )

    class UnsignedNotAllowed :
        SignatureVerificationException(
            "Unsigned bundle not allowed when signing is enabled. " +
                "Public key is configured but bundle is not signed. Rejecting update",
        )

    class SecurityFrameworkError(
        cause: Throwable,
    ) : SignatureVerificationException(
            "Security framework error during verification: ${cause.message}",
        )
}

/**
 * Service for verifying bundle integrity through hash or RSA-SHA256 signature verification.
 * Uses Java Signature API for cryptographic operations.
 *
 * fileHash format:
 * - Signed: `sig:<base64_signature>` - Verify signature (implicitly verifies hash)
 * - Unsigned: `<hex_hash>` - Verify SHA256 hash only
 *
 * Security rules:
 * - null/empty fileHash → REJECT
 * - sig:... + public key configured → verify signature → Install/REJECT
 * - sig:... + public key NOT configured → REJECT (can't verify)
 * - <hash> + public key configured → REJECT (unsigned not allowed)
 * - <hash> + public key NOT configured → verify hash → Install/REJECT
 */
object SignatureVerifier {
    private const val TAG = "SignatureVerifier"

    /**
     * Reads public key from Android string resources.
     * @param context Application context
     * @return Public key PEM string or null if not configured
     */
    private fun getPublicKeyFromConfig(context: Context): String? {
        val resourceId = StringResourceUtils.getIdentifier(context, "hot_updater_public_key")

        if (resourceId == 0) {
            Log.d(TAG, "hot_updater_public_key not found in strings.xml")
            return null
        }

        val publicKeyPEM = context.getString(resourceId)
        if (publicKeyPEM.isEmpty()) {
            Log.d(TAG, "hot_updater_public_key is empty")
            return null
        }

        return publicKeyPEM
    }

    /**
     * Checks if signing is enabled (public key is configured).
     * @param context Application context
     * @return true if public key is configured
     */
    fun isSigningEnabled(context: Context): Boolean = getPublicKeyFromConfig(context) != null

    /**
     * Checks if fileHash is in signed format (starts with "sig:").
     * @param fileHash The file hash string to check
     * @return true if signed format
     */
    fun isSignedFormat(fileHash: String?): Boolean = fileHash?.startsWith(SIGNED_HASH_PREFIX) == true

    /**
     * Extracts signature from signed format fileHash.
     * @param fileHash The signed file hash (sig:<signature>)
     * @return Base64-encoded signature or null if not signed format
     */
    fun extractSignature(fileHash: String?): String? {
        if (!isSignedFormat(fileHash)) return null
        return fileHash?.removePrefix(SIGNED_HASH_PREFIX)
    }

    /**
     * Verifies bundle integrity based on fileHash format.
     * Determines verification mode by checking for "sig:" prefix.
     *
     * @param context Application context
     * @param bundleFile The bundle file to verify
     * @param fileHash Combined hash string (sig:<signature> or <hex_hash>)
     * @throws SignatureVerificationException if verification fails
     */
    fun verifyBundle(
        context: Context,
        bundleFile: File,
        fileHash: String?,
    ) {
        val signingEnabled = isSigningEnabled(context)

        // Rule: null/empty fileHash → REJECT
        if (fileHash.isNullOrEmpty()) {
            Log.e(TAG, "fileHash is null or empty. Rejecting update.")
            throw SignatureVerificationException.MissingFileHash()
        }

        if (isSignedFormat(fileHash)) {
            // Signed format: sig:<signature>
            val signature = extractSignature(fileHash)
            if (signature.isNullOrEmpty()) {
                Log.e(TAG, "Failed to extract signature from fileHash")
                throw SignatureVerificationException.InvalidSignatureFormat()
            }

            // Rule: sig:... + public key NOT configured → REJECT
            if (!signingEnabled) {
                Log.e(TAG, "Signed bundle but public key not configured. Cannot verify.")
                throw SignatureVerificationException.PublicKeyNotConfigured()
            }

            // Rule: sig:... + public key configured → verify signature
            verifySignature(context, bundleFile, signature)
        } else {
            // Unsigned format: <hex_hash>

            // Rule: <hash> + public key configured → REJECT
            if (signingEnabled) {
                Log.e(TAG, "Unsigned bundle not allowed when signing is enabled. Rejecting.")
                throw SignatureVerificationException.UnsignedNotAllowed()
            }

            // Rule: <hash> + public key NOT configured → verify hash
            verifyHash(bundleFile, fileHash)
        }
    }

    /**
     * Verifies SHA256 hash of a file.
     * @param bundleFile The file to verify
     * @param expectedHash Expected SHA256 hash (hex string)
     * @throws SignatureVerificationException.FileHashMismatch if verification fails
     */
    fun verifyHash(
        bundleFile: File,
        expectedHash: String,
    ) {
        Log.d(TAG, "Verifying hash for file: ${bundleFile.name}")

        if (!HashUtils.verifyHash(bundleFile, expectedHash)) {
            Log.e(TAG, "Hash mismatch!")
            throw SignatureVerificationException.FileHashMismatch()
        }

        Log.i(TAG, "✅ Hash verified successfully")
    }

    /**
     * Verifies RSA-SHA256 signature of a file.
     * Calculates the file hash internally and verifies the signature.
     *
     * @param context Application context
     * @param bundleFile The file to verify
     * @param signatureBase64 Base64-encoded RSA-SHA256 signature
     * @throws SignatureVerificationException if verification fails
     */
    fun verifySignature(
        context: Context,
        bundleFile: File,
        signatureBase64: String,
    ) {
        Log.d(TAG, "Verifying signature for file: ${bundleFile.name}")

        // Get public key from config
        val publicKeyPEM =
            getPublicKeyFromConfig(context)
                ?: run {
                    Log.e(TAG, "Cannot verify signature: public key not configured in strings.xml")
                    throw SignatureVerificationException.PublicKeyNotConfigured()
                }

        try {
            // Convert PEM to PublicKey
            val publicKey = createPublicKey(publicKeyPEM)

            // Calculate file hash
            val fileHashHex =
                HashUtils.calculateSHA256(bundleFile)
                    ?: run {
                        Log.e(TAG, "Failed to calculate file hash")
                        throw SignatureVerificationException.FileReadFailed()
                    }

            Log.d(TAG, "Calculated file hash: $fileHashHex")

            // Decode signature from base64
            val signatureBytes =
                try {
                    Base64.decode(signatureBase64, Base64.DEFAULT)
                } catch (e: Exception) {
                    Log.e(TAG, "Failed to decode signature from base64", e)
                    throw SignatureVerificationException.InvalidSignatureFormat()
                }

            // Convert hex fileHash to bytes
            val fileHashBytes = hexToByteArray(fileHashHex)

            // Verify signature using RSA-SHA256
            val verifier = Signature.getInstance("SHA256withRSA")
            verifier.initVerify(publicKey)
            verifier.update(fileHashBytes)
            val isValid = verifier.verify(signatureBytes)

            if (isValid) {
                Log.i(TAG, "✅ Signature verified successfully")
            } else {
                Log.e(TAG, "❌ Signature verification failed")
                throw SignatureVerificationException.SignatureVerificationFailed()
            }
        } catch (e: SignatureVerificationException) {
            throw e
        } catch (e: Exception) {
            Log.e(TAG, "Signature verification error", e)
            throw SignatureVerificationException.SecurityFrameworkError(e)
        }
    }

    /**
     * Converts PEM-formatted public key to PublicKey.
     * @param publicKeyPEM Public key in PEM format
     * @return PublicKey instance
     * @throws SignatureVerificationException.InvalidPublicKeyFormat if conversion fails
     */
    private fun createPublicKey(publicKeyPEM: String): PublicKey {
        try {
            // Remove PEM headers/footers and whitespace
            val publicKeyBase64 =
                publicKeyPEM
                    .replace("-----BEGIN PUBLIC KEY-----", "")
                    .replace("-----END PUBLIC KEY-----", "")
                    .replace("\\n", "")
                    .replace("\n", "")
                    .replace("\r", "")
                    .replace(" ", "")
                    .trim()

            // Decode base64
            val keyBytes = Base64.decode(publicKeyBase64, Base64.DEFAULT)

            // Create PublicKey from X.509 format (SubjectPublicKeyInfo)
            val spec = X509EncodedKeySpec(keyBytes)
            val keyFactory = KeyFactory.getInstance("RSA")
            val publicKey = keyFactory.generatePublic(spec)

            Log.d(TAG, "Public key loaded successfully")
            return publicKey
        } catch (e: Exception) {
            Log.e(TAG, "Failed to create public key", e)
            throw SignatureVerificationException.InvalidPublicKeyFormat()
        }
    }

    /**
     * Converts hex string to ByteArray.
     * @param hexString Hex-encoded string
     * @return ByteArray
     * @throws SignatureVerificationException.SignatureVerificationFailed if conversion fails
     */
    private fun hexToByteArray(hexString: String): ByteArray {
        try {
            val len = hexString.length
            if (len % 2 != 0) {
                throw SignatureVerificationException.InvalidSignatureFormat()
            }

            val data = ByteArray(len / 2)
            var i = 0
            while (i < len) {
                data[i / 2] =
                    (
                        (Character.digit(hexString[i], 16) shl 4) +
                            Character.digit(hexString[i + 1], 16)
                    ).toByte()
                i += 2
            }
            return data
        } catch (e: Exception) {
            Log.e(TAG, "Failed to convert hex to byte array", e)
            throw SignatureVerificationException.InvalidSignatureFormat()
        }
    }
}

File: packages/react-native/ios/HotUpdater/Internal/SignatureVerifier.swift
import Foundation
import Security

/// Prefix for signed file hash format.
private let SIGNED_HASH_PREFIX = "sig:"

/// Error types for signature verification failures.
///
/// **IMPORTANT**: The error messages in `errorUserInfo` are used by the JavaScript layer
/// (`packages/react-native/src/types.ts`) to detect signature verification failures.
/// If you change these messages, update `isSignatureVerificationError()` in types.ts accordingly.
public enum SignatureVerificationError: Error, CustomNSError {
    case publicKeyNotConfigured
    case invalidPublicKeyFormat
    case invalidSignatureFormat
    case verificationFailed
    case hashMismatch
    case hashCalculationFailed
    case unsignedNotAllowed
    case securityFrameworkError(OSStatus)

    // CustomNSError protocol implementation
    public static var errorDomain: String {
        return "com.hotupdater.SignatureVerificationError"
    }

    public var errorCode: Int {
        switch self {
        case .publicKeyNotConfigured: return 2001
        case .invalidPublicKeyFormat: return 2002
        case .invalidSignatureFormat: return 2003
        case .verificationFailed: return 2004
        case .hashMismatch: return 2005
        case .hashCalculationFailed: return 2006
        case .unsignedNotAllowed: return 2007
        case .securityFrameworkError: return 2099
        }
    }

    public var errorUserInfo: [String: Any] {
        var userInfo: [String: Any] = [:]

        switch self {
        case .publicKeyNotConfigured:
            userInfo[NSLocalizedDescriptionKey] = "Public key not configured for signature verification"
            userInfo[NSLocalizedRecoverySuggestionErrorKey] = "Add HOT_UPDATER_PUBLIC_KEY to Info.plist with your RSA public key"

        case .invalidPublicKeyFormat:
            userInfo[NSLocalizedDescriptionKey] = "Public key format is invalid"
            userInfo[NSLocalizedRecoverySuggestionErrorKey] = "Ensure the public key is in PEM format (BEGIN PUBLIC KEY)"

        case .invalidSignatureFormat:
            userInfo[NSLocalizedDescriptionKey] = "Signature format is invalid"
            userInfo[NSLocalizedRecoverySuggestionErrorKey] = "The signature must be base64-encoded"

        case .verificationFailed:
            userInfo[NSLocalizedDescriptionKey] = "Bundle signature verification failed"
            userInfo[NSLocalizedRecoverySuggestionErrorKey] = "The bundle may be corrupted or tampered with. Rejecting update for security"

        case .hashMismatch:
            userInfo[NSLocalizedDescriptionKey] = "Bundle hash verification failed"
            userInfo[NSLocalizedRecoverySuggestionErrorKey] = "The bundle file hash does not match. File may be corrupted"

        case .hashCalculationFailed:
            userInfo[NSLocalizedDescriptionKey] = "Failed to calculate file hash"
            userInfo[NSLocalizedRecoverySuggestionErrorKey] = "Could not read file for hash verification"

        case .unsignedNotAllowed:
            userInfo[NSLocalizedDescriptionKey] = "Unsigned bundle not allowed when signing is enabled"
            userInfo[NSLocalizedRecoverySuggestionErrorKey] = "Public key is configured but bundle is not signed. Rejecting update"

        case .securityFrameworkError(let status):
            userInfo[NSLocalizedDescriptionKey] = "Security framework error during verification (OSStatus: \(status))"
            userInfo[NSLocalizedRecoverySuggestionErrorKey] = "Check public key format and signature data"
        }

        return userInfo
    }
}

/**
 * Service for verifying bundle integrity through hash or RSA-SHA256 signature verification.
 * Uses iOS Security framework for cryptographic operations.
 *
 * fileHash format:
 * - Signed: `sig:<base64_signature>` - Verify signature (implicitly verifies hash)
 * - Unsigned: `<hex_hash>` - Verify SHA256 hash only
 *
 * Security rules:
 * - null/empty fileHash → REJECT
 * - sig:... + public key configured → verify signature → Install/REJECT
 * - sig:... + public key NOT configured → REJECT (can't verify)
 * - <hash> + public key configured → REJECT (unsigned not allowed)
 * - <hash> + public key NOT configured → verify hash → Install/REJECT
 */
public class SignatureVerifier {

    /**
     * Reads public key from Info.plist configuration.
     * @return Public key PEM string or nil if not configured
     */
    private static func getPublicKeyFromConfig() -> String? {
        guard let publicKeyPEM = Bundle.main.object(forInfoDictionaryKey: "HOT_UPDATER_PUBLIC_KEY") as? String else {
            NSLog("[SignatureVerifier] HOT_UPDATER_PUBLIC_KEY not found in Info.plist")
            return nil
        }
        return publicKeyPEM
    }

    /**
     * Checks if signing is enabled (public key is configured).
     * @return true if public key is configured
     */
    public static func isSigningEnabled() -> Bool {
        return getPublicKeyFromConfig() != nil
    }

    /**
     * Checks if fileHash is in signed format (starts with "sig:").
     * @param fileHash The file hash string to check
     * @return true if signed format
     */
    public static func isSignedFormat(_ fileHash: String?) -> Bool {
        guard let hash = fileHash else { return false }
        return hash.hasPrefix(SIGNED_HASH_PREFIX)
    }

    /**
     * Extracts signature from signed format fileHash.
     * @param fileHash The signed file hash (sig:<signature>)
     * @return Base64-encoded signature or nil if not signed format
     */
    public static func extractSignature(_ fileHash: String?) -> String? {
        guard let hash = fileHash, isSignedFormat(hash) else { return nil }
        return String(hash.dropFirst(SIGNED_HASH_PREFIX.count))
    }

    /**
     * Verifies bundle integrity based on fileHash format.
     * Determines verification mode by checking for "sig:" prefix.
     *
     * @param fileURL URL of the bundle file to verify
     * @param fileHash Combined hash string (sig:<signature> or <hex_hash>)
     * @return Result indicating verification success or failure with error
     */
    public static func verifyBundle(fileURL: URL, fileHash: String?) -> Result<Void, SignatureVerificationError> {
        let signingEnabled = isSigningEnabled()

        // Rule: null/empty fileHash → REJECT
        guard let hash = fileHash, !hash.isEmpty else {
            NSLog("[SignatureVerifier] fileHash is null or empty. Rejecting update.")
            return .failure(.verificationFailed)
        }

        if isSignedFormat(hash) {
            // Signed format: sig:<signature>
            guard let signature = extractSignature(hash) else {
                NSLog("[SignatureVerifier] Failed to extract signature from fileHash")
                return .failure(.invalidSignatureFormat)
            }

            // Rule: sig:... + public key NOT configured → REJECT
            guard signingEnabled else {
                NSLog("[SignatureVerifier] Signed bundle but public key not configured. Cannot verify.")
                return .failure(.publicKeyNotConfigured)
            }

            // Rule: sig:... + public key configured → verify signature
            return verifySignature(fileURL: fileURL, signatureBase64: signature)
        } else {
            // Unsigned format: <hex_hash>

            // Rule: <hash> + public key configured → REJECT
            if signingEnabled {
                NSLog("[SignatureVerifier] Unsigned bundle not allowed when signing is enabled. Rejecting.")
                return .failure(.unsignedNotAllowed)
            }

            // Rule: <hash> + public key NOT configured → verify hash
            return verifyHash(fileURL: fileURL, expectedHash: hash)
        }
    }

    /**
     * Verifies SHA256 hash of a file.
     * @param fileURL URL of the file to verify
     * @param expectedHash Expected SHA256 hash (hex string)
     * @return Result indicating verification success or failure
     */
    public static func verifyHash(fileURL: URL, expectedHash: String) -> Result<Void, SignatureVerificationError> {
        NSLog("[SignatureVerifier] Verifying hash for file: \(fileURL.lastPathComponent)")

        guard HashUtils.verifyHash(fileURL: fileURL, expectedHash: expectedHash) else {
            NSLog("[SignatureVerifier] Hash mismatch!")
            return .failure(.hashMismatch)
        }

        NSLog("[SignatureVerifier] ✅ Hash verified successfully")
        return .success(())
    }

    /**
     * Verifies RSA-SHA256 signature of a file.
     * Calculates the file hash internally and verifies the signature.
     *
     * @param fileURL URL of the file to verify
     * @param signatureBase64 Base64-encoded RSA-SHA256 signature
     * @return Result indicating verification success or failure with error
     */
    public static func verifySignature(fileURL: URL, signatureBase64: String) -> Result<Void, SignatureVerificationError> {
        NSLog("[SignatureVerifier] Verifying signature for file: \(fileURL.lastPathComponent)")

        // Get public key from config
        guard let publicKeyPEM = getPublicKeyFromConfig() else {
            NSLog("[SignatureVerifier] Cannot verify signature: public key not configured in Info.plist")
            return .failure(.publicKeyNotConfigured)
        }

        // Convert PEM to SecKey
        let publicKeyResult = createPublicKey(from: publicKeyPEM)
        guard case .success(let publicKey) = publicKeyResult else {
            if case .failure(let error) = publicKeyResult {
                return .failure(error)
            }
            return .failure(.invalidPublicKeyFormat)
        }

        // Calculate file hash
        guard let fileHashHex = HashUtils.calculateSHA256(fileURL: fileURL) else {
            NSLog("[SignatureVerifier] Failed to calculate file hash")
            return .failure(.hashCalculationFailed)
        }

        NSLog("[SignatureVerifier] Calculated file hash: \(fileHashHex)")

        // Decode signature from base64
        guard let signatureData = Data(base64Encoded: signatureBase64) else {
            NSLog("[SignatureVerifier] Failed to decode signature from base64")
            return .failure(.invalidSignatureFormat)
        }

        // Convert hex fileHash to data
        guard let fileHashData = dataFromHexString(fileHashHex) else {
            NSLog("[SignatureVerifier] Failed to convert fileHash from hex")
            return .failure(.invalidSignatureFormat)
        }

        // Verify signature
        let algorithm: SecKeyAlgorithm = .rsaSignatureMessagePKCS1v15SHA256

        guard SecKeyIsAlgorithmSupported(publicKey, .verify, algorithm) else {
            NSLog("[SignatureVerifier] RSA-SHA256 algorithm not supported")
            return .failure(.securityFrameworkError(-1))
        }

        var error: Unmanaged<CFError>?
        let verified = SecKeyVerifySignature(
            publicKey,
            algorithm,
            fileHashData as CFData,
            signatureData as CFData,
            &error
        )

        if let err = error?.takeRetainedValue() {
            NSLog("[SignatureVerifier] Verification failed: \(err)")
            return .failure(.verificationFailed)
        }

        if verified {
            NSLog("[SignatureVerifier] ✅ Signature verified successfully")
            return .success(())
        } else {
            NSLog("[SignatureVerifier] ❌ Signature verification failed")
            return .failure(.verificationFailed)
        }
    }

    /**
     * Converts PEM-formatted public key to SecKey.
     * @param publicKeyPEM Public key in PEM format
     * @return SecKey or error
     */
    private static func createPublicKey(from publicKeyPEM: String) -> Result<SecKey, SignatureVerificationError> {
        // Remove PEM headers/footers and whitespace
        var keyString = publicKeyPEM
            .replacingOccurrences(of: "-----BEGIN PUBLIC KEY-----", with: "")
            .replacingOccurrences(of: "-----END PUBLIC KEY-----", with: "")
            .replacingOccurrences(of: "\\n", with: "")
            .replacingOccurrences(of: "\n", with: "")
            .replacingOccurrences(of: " ", with: "")

        // Decode base64
        guard let keyData = Data(base64Encoded: keyString) else {
            NSLog("[SignatureVerifier] Failed to decode public key from base64")
            return .failure(.invalidPublicKeyFormat)
        }

        // SecKeyCreateWithData auto-detects key size from SPKI-formatted key data.
        // This supports any valid RSA key size (2048, 3072, 4096-bit, etc.)
        let attributes: [String: Any] = [
            kSecAttrKeyType as String: kSecAttrKeyTypeRSA,
            kSecAttrKeyClass as String: kSecAttrKeyClassPublic
        ]

        var error: Unmanaged<CFError>?
        guard let secKey = SecKeyCreateWithData(keyData as CFData, attributes as CFDictionary, &error) else {
            if let err = error?.takeRetainedValue() {
                NSLog("[SignatureVerifier] SecKeyCreateWithData failed: \(err)")
            }
            return .failure(.invalidPublicKeyFormat)
        }

        return .success(secKey)
    }

    /**
     * Converts hex string to Data.
     * @param hexString Hex-encoded string
     * @return Data or nil if invalid format
     */
    private static func dataFromHexString(_ hexString: String) -> Data? {
        var data = Data(capacity: hexString.count / 2)

        var index = hexString.startIndex
        while index < hexString.endIndex {
            let nextIndex = hexString.index(index, offsetBy: 2)
            guard nextIndex <= hexString.endIndex else { return nil }

            let byteString = hexString[index..<nextIndex]
            guard let byte = UInt8(byteString, radix: 16) else { return nil }

            data.append(byte)
            index = nextIndex
        }

        return data
    }
}
